web/tests/compose_ui.test.js

```diff
@@ -1273,6 +1273,93 @@ run_test("format_text - spoiler", ({override}) => {
     assert.equal(wrap_selection_called, false);
 });
 
+run_test("format_text - link", ({override}) => {
+    override(text_field_edit, "set", (_field, text) => {
+        set_text = text;
+    });
+    override(text_field_edit, "wrapSelection", (_field, syntax_start, syntax_end) => {
+        wrap_selection_called = true;
+        wrap_syntax_start = syntax_start;
+        wrap_syntax_end = syntax_end;
+    });
+
+    const link_syntax_start = "[";
+    const link_syntax_end = "](url)";
+
+    // Link selected text
+    reset_state();
+    init_textarea("abc", {
+        start: 0,
+        end: 3,
+        text: "abc",
+        length: 3,
+    });
+    compose_ui.format_text($textarea, "link");
+    assert.equal(set_text, "");
+    assert.equal(wrap_selection_called, true);
+    assert.equal(wrap_syntax_start, link_syntax_start);
+    assert.equal(wrap_syntax_end, link_syntax_end);
+
+    // Undo link selected text, url selected
+    reset_state();
+    init_textarea("[abc](def)", {
+        start: 6,
+        end: 9,
+        text: "def",
+        length: 3,
+    });
+    compose_ui.format_text($textarea, "link");
+    assert.equal(set_text, "abc def");
+    assert.equal(wrap_selection_called, false);
+
+    reset_state();
+    init_textarea("[abc](url)", {
+        start: 6,
+        end: 9,
+        text: "url",
+        length: 3,
+    });
+    compose_ui.format_text($textarea, "link");
+    assert.equal(set_text, "abc");
+    assert.equal(wrap_selection_called, false);
```

Comment:
We should really rework this to have a way cleaner way to express these tests with way less verbose, by-hand logic; in theory a test with a cursor in a given position could be expressed as `textBeforeCursor|textAfterCursor` with `|` as the cursor, and maybe `textBeforeSelection<selection>textAfterSelection` in the tests, and have a test helper be responsible for everything else.

So this test would become something like this, while verifying the same set of facts about behavior:

assert.equal(test_compose_button("link", "[abc](<def>)"), "abc|");

I think that's basically a follow-up PR though.

Reply:
Here's the follow up PR revamping these tests: #28554 (also has 2 commits fixing a bug each that the new tests helped surface)